### DISCUSSION TITLE
return new Response seems to send headers that disrupts PDF rendering as part of response stream

### DISCUSSION THREAD

**Discussion by hornlp** [symfony/symfony#54575] [Q&A]
I recently upgraded my project using composer (minor upgrade in Symfony 6.x). My rendering of PDF using DOMPDF was working, but now getting a "Headers already sent" error in log and not rendering the PDF. I see headers are set at **new Response**. Is there a way to get this to work again?

```
return new Response (
        $dompdf->stream($filename, ["Attachment" => false]),
        Response::HTTP_OK, ['Content-Type' => 'application/pdf']
 );
```

If I use ...

```
$response = new Response (
        $dompdf->stream($filename, ["Attachment" => false]),
        Response::HTTP_OK, ['Content-Type' => 'application/pdf']
);

return $response->send();
```

It Works! I get the PDF response. However, still get the "Headers already sent" error in log.

`Uncaught PHP Exception ErrorException: "Warning: Cannot modify header information - headers already sent by (output started at /{path-to-project}/vendor/dompdf/dompdf/src/Adapter/CPDF.php:922)" at SessionUtils.php line 5`

**Comment by alexandre-le-borgne** [comment]
Please add the code related to the error

CPDF.php:922 and around

**Reply by hornlp** [reply]
```
public function stream($filename = "document.pdf", $options = [])
    {
        if (headers_sent()) {
            die("Unable to stream pdf: headers already sent");
        }

        if (!isset($options["compress"])) $options["compress"] = true;
        if (!isset($options["Attachment"])) $options["Attachment"] = true;

        $debug = !$options['compress'];
        $tmp = ltrim($this->_pdf->output($debug));

        header("Cache-Control: private");
        header("Content-Type: application/pdf");
        header("Content-Length: " . mb_strlen($tmp, "8bit"));

        $filename = str_replace(["\n", "'"], "", basename($filename, ".pdf")) . ".pdf";
        $attachment = $options["Attachment"] ? "attachment" : "inline";
        header(Helpers::buildContentDispositionHeader($attachment, $filename));

        echo $tmp;
        flush();
    }

    public function output($options = [])
    {
        if (!isset($options["compress"])) $options["compress"] = true;

        $debug = !$options['compress'];

        return $this->_pdf->output($debug);
    }
```

**Comment by smnandre** [comment]
You should try with `$dompdf->output(...)` instead of `$dompdf->stream(...)`

**Reply by hornlp** [reply]
Same result with `$dompdf->output(...)`. Get PDF, but similar error in log.

`[Application] Apr 13 07:58:30 |CRITICA| REQUES Uncaught PHP Exception ErrorException: "Warning: Cannot modify header information - headers already sent by (output started at {path-to-project}/vendor/symfony/http-foundation/Response.php:408)" at SessionUtils.php line 52`

**Reply by hornlp** [reply]
However this time indicating it at **Response.php:408**.

```
    public function sendContent(): static
    {
        echo $this->content;

        return $this;
    }
```

**Comment by hornlp** [comment]
It seems that on production I do not have this error in the log. Could it be from the Web Debug Toolbar or something else on Dev?
